protocol/msgpack/raw_metric.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/aggregated"
+)
+
+var (
+	emptyMetric aggregated.Metric
+)
+
+type decodeVersionFn func() int
+type decodeBytesLenFn func() int
+type decodeTimeFn func() time.Time
+type decodeFloat64Fn func() float64
+type readBytesFn func(n int) []byte
+
+// rawMetric is a raw metric
+type rawMetric struct {
+	data             []byte            // raw data containing encoded metric
+	buf              *bytes.Buffer     // intermediate buffer
+	it               iteratorBase      // base iterator for lazily decoding metric fields
+	metric           aggregated.Metric // current metric
+	idDecoded        bool              // whether id has been decoded
+	timestampDecoded bool              // whether timestamp has been decoded
+	valueDecoded     bool              // whether value has been decoded
+	decodeVersionFn  decodeVersionFn   // decoding version function
+	decodeBytesLenFn decodeBytesLenFn  // decoding the byte slice length function
+	decodeTimeFn     decodeTimeFn      // decoding time function
+	decodeFloat64Fn  decodeFloat64Fn   // decoding float64 value function
+	readBytesFn      readBytesFn       // reading bytes function
```

Comment:
Are all these fns super necessary for testing?  It adds a large overhead to `rawMetric` which could be a very high cardinality object.  5 * 8 = extra 40 bytes per metric.